kernel/src/schema_compat.rs

```diff
@@ -0,0 +1,331 @@
+            StructField::new("name", DataType::STRING, true),
+            StructField::new("age", DataType::INTEGER, true),
+        ]);
+        assert!(existing_schema.can_read_as(&read_schema).is_ok());
+    }
+    #[test]
+    fn set_nullable_to_false_fails() {
+        let existing_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, true),
+        ]);
+        let read_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, false),
+        ]);
+        assert!(existing_schema.can_read_as(&read_schema).is_err());
+    }
+    #[test]
+    fn new_nullable_column() {
+        let existing_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, true),
+        ]);
+
+        let read_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, true),
+            StructField::new("location", DataType::STRING, true),
+        ]);
+        assert!(existing_schema.can_read_as(&read_schema).is_ok());
+    }
+    #[test]
+    fn new_non_nullable_column_fails() {
+        let existing_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, true),
+        ]);
+
+        let read_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, true),
+            StructField::new("location", DataType::STRING, false),
+        ]);
+        assert!(existing_schema.can_read_as(&read_schema).is_err());
+    }
+    #[test]
+    fn duplicate_field_modulo_case() {
+        let existing_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("Id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, true),
+        ]);
+
+        let read_schema = StructType::new([
+            StructField::new("id", DataType::LONG, false),
+            StructField::new("Id", DataType::LONG, false),
+            StructField::new("name", DataType::STRING, false),
+            StructField::new("age", DataType::INTEGER, true),
+        ]);
+        assert!(existing_schema.can_read_as(&read_schema).is_err());
+    }
```

Comment:
can we add a test for when the read field names aren't a subset of the original Schema names?
e.g. read `{a, b, c}` for struct/schema `{a, b}`?

Reply:
Worked this in the new `differ_by_nullable_column` and `differ_by_non_nullable_column`. It checks both directions.